File: Difficulty: Easy/Longest Common Prefix of Strings/longest-common-prefix-of-strings.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++

class Solution {
  public:

    string longestCommonPrefix(vector<string> arr) {
        // your code here
        int small = 0;
        for(int i=0;i<arr.size();i++){
            if(arr[i].length() < arr[small].length()){
                small = i;
            }
        }
        string temp = arr[small];
        
        int i=0,j=0;
        while(i<temp.length()){
            for(j=0;j<arr.size();j++){
                if(temp[i] != arr[j][i]){
                    break;
                }
            }
            if(j != arr.size()) break;
            i++;
        }
        if(i==0) return "-1";
        return temp.substr(0,i);
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        vector<string> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        string number;
        while (ss >> number) {
            arr.push_back(number);
        }

        Solution ob;
        cout << ob.longestCommonPrefix(arr) << endl;
    }
}

// } Driver Code Ends

File: Difficulty: Easy/Longest Common Prefix of Strings/longest-common-prefix-of-strings.js
//{ Driver Code Starts
// Initial Template for javascript

"use strict";

process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;

process.stdin.on("data", (inputStdin) => { inputString += inputStdin; });

process.stdin.on("end", (_) => {
    inputString =
        inputString.trim().split("\n").map((string) => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

/* Function to print an array */
function printArray(arr, size) {
    let i;
    let s = "";
    for (i = 0; i < size; i++) {
        s += arr[i] + " ";
    }
    console.log(s);
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {
        let arr = readLine().split(" ");
        const ob = new Solution();
        const ans = ob.longestCommonPrefix(arr);

        if (ans === "") {
            console.log("\"\"");
        } else
            console.log(ans);
    }
}
// } Driver Code Ends


// User function Template for javascript
/**
 * @param {string[]} arr
 * @param {number} n
 * @returns {string}
 */
class Solution {
    longestCommonPrefix(arr) {
        // code here
        if(arr.length===0) return "";
        
        if(arr.length===1) return arr[0];
        
        arr.sort();
        
        let first = arr[0];
        let last = arr[arr.length-1];
        let i=0;
        
        let minLen = Math.min(first.length, last.length);
        
       while(i<minLen && first[i]===last[i]){
           i++;
       }
        
        return first.substring(0,i);
    }
}